commands/operator-sdk/cmd/generate/openapi.go

```diff
@@ -0,0 +1,173 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func NewGenerateOpenAPICmd() *cobra.Command {
+	openAPICmd := &cobra.Command{
+		Use:   "openapi",
+		Short: "Generates OpenAPI specs for API's",
+		Long: ` generate openapi generates OpenAPI validation specs in Go from types in
+all pkg/apis/<group>/<version> directories.
+`,
+		RunE: openAPIFunc,
+	}
+
+	return openAPICmd
+}
+
+func openAPIFunc(cmd *cobra.Command, args []string) error {
+	if len(args) != 0 {
+		return fmt.Errorf("command %s doesn't accept any arguments", cmd.CommandPath())
+	}
+
+	return OpenAPIGen()
+}
+
+// OpenAPIGen generates OpenAPI validation specs for all CRD's in dirs.
+func OpenAPIGen() error {
+	projutil.MustInProjectRoot()
+
+	absProjectPath := projutil.MustGetwd()
+	repoPkg := projutil.CheckAndGetProjectGoPkg()
+	vendor := filepath.Join(absProjectPath, "vendor")
+	srcDir := filepath.Join(vendor, "k8s.io", "kube-openapi")
+	bpFile := filepath.Join(vendor, "k8s.io", "gengo", "boilerplate", "boilerplate.go.txt")
+	binDir := filepath.Join(absProjectPath, scaffold.BuildBinDir)
+
+	if err := buildOpenAPIGenBinary(binDir, srcDir); err != nil {
+		return err
+	}
+
+	gvMap, err := genutil.ParseGroupVersions()
+	if err != nil {
+		return fmt.Errorf("failed to parse group versions: (%v)", err)
+	}
+	gvb := &strings.Builder{}
+	for g, vs := range gvMap {
+		gvb.WriteString(fmt.Sprintf("%s:%v, ", g, vs))
+	}
+
+	log.Infof("Running OpenAPI code-generation for Custom Resource group versions: [%v]\n", gvb.String())
+
+	apisPkg := filepath.Join(repoPkg, scaffold.ApisDir)
+	fqApiStr := genutil.CreateFQApis(apisPkg, gvMap)
+	fqApis := strings.Split(fqApiStr, ",")
+	if err := openAPIGen(binDir, bpFile, fqApis); err != nil {
+		return err
+	}
+
+	s := &scaffold.Scaffold{}
+	cfg := &input.Config{
+		Repo:           repoPkg,
+		AbsProjectPath: absProjectPath,
+		ProjectName:    filepath.Base(absProjectPath),
+	}
+	crdMap, err := getCRDGVKMap()
+	if err != nil {
+		return err
+	}
+	for g, vs := range gvMap {
+		for _, v := range vs {
+			gvks := crdMap[filepath.Join(g, v)]
+			for _, gvk := range gvks {
+				r, err := scaffold.NewResource(filepath.Join(gvk.Group, gvk.Version), gvk.Kind)
+				if err != nil {
+					return err
+				}
+				err = s.Execute(cfg,
+					&scaffold.CRD{Resource: r, IsOperatorGo: projutil.IsOperatorGo()},
+				)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	log.Info("Code-generation complete.")
+	return nil
+}
+
+func buildOpenAPIGenBinary(binDir, codegenSrcDir string) error {
+	genDirs := []string{"./cmd/openapi-gen"}
+	return genutil.BuildCodegenBinaries(genDirs, binDir, codegenSrcDir)
+}
+
+func openAPIGen(binDir, headerFile string, fqApis []string) error {
+	cgPath := filepath.Join(binDir, "openapi-gen")
+	for _, fqApi := range fqApis {
+		args := []string{
+			"--input-dirs", fqApi,
+			"--output-package", fqApi,
+			"--output-file-base", "zz_generated.openapi",
+			"--go-header-file", headerFile,
```

Comment:
Just a nit but I don't think we should be adding the [Kubernetes header license](https://github.com/kubernetes/gengo/blob/master/boilerplate/boilerplate.go.txt) by default into the generated files. We don't do that for any other files in the project.
So I would vote to keep this out and not pass in the `--go-header-file` flag. Same for deepcopy.

Reply:
Same here. Earlier the generator would throw an error if this flag was not set, but it does not now for some reason. Removing.